File: src/config_manager/ConfigManager.cpp
#include "ConfigManager.h"
#include <stdexcept>

ConfigManager::ConfigManager(const std::string& configPath)
    : _configPath(configPath)
{
    loadConfig();
}

ConfigManager& ConfigManager::getInstance(const std::string& configPath)
{
    static ConfigManager instance(configPath);
    return instance;
}

void ConfigManager::loadConfig()
{
    try {
        _config = YAML::LoadFile(_configPath);
    } catch (const std::exception& e) {
        throw std::runtime_error(std::string("Error loading config: ") + e.what());
    }
}

std::string ConfigManager::get(const std::string& key, const std::string& defaultValue) const
{
    if (_config[key]) {
        return _config[key].as<std::string>();
    }
    return defaultValue;
}

YAML::Node ConfigManager::getSection(const std::string& section) const
{
    if (_config[section]) {
        return _config[section];
    }
    return YAML::Node();
}

File: archive/src_old_ib_connection/config_manager/config_manager.py
import yaml

class ConfigManager:
    _instance = None

    ## Singleton config
    def __new__(cls, config_path='config/global_config.yaml'):
        if cls._instance is None:
            cls._instance = super(ConfigManager, cls).__new__(cls)
            cls._instance._config_path = config_path
            cls._instance._config = {}
            cls._instance.load_config()
        return cls._instance
    
    def load_config(self):
        try:
            with open(self._config_path, 'r') as f:
                self._config = yaml.safe_load(f)
        except Exception as e:
            raise RuntimeError(f"Error loading config: {e}")

    ## Getter  
    def get(self,key,default=None):
        return self._config.get(key,default)
    
    def get_section(self,section):
        return self._config.get(section,{})
